docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md

```diff
@@ -0,0 +1,114 @@
+---
+title: Setup a Highly Available Etcd Cluster With Kubeadm
+---
+
+{% capture overview %}
+Kubeadm by default will create a self-hosted etcd cluster of one member. This is not a great default for a highly available control plane. This task walks through the process of creating a highly available etcd cluster of three members that can be used as an external etcd when using kubeadm to set up a kubernetes cluster.
+{% endcapture %}
+
+{% capture prerequisites %}
+* Three hosts that can talk to each other over default etcd ports (2379 for clients and 2380 for peers).
+* The `kubeadm` binary installed on each host.
+* SSH or similar to one host and from that host, SSH needs to work to both other hosts (to copy certificate files around).
+* Docker installed on every host.
+* The `kubelet` installed and running with configuration to read static manifests from `/etc/kubernetes/manifests/`.
+{% endcapture %}
+
+{% capture steps %}
+
+## Kubeadm Configuration File
+
+Create a kubeadm configuration file on each host that looks something like this:
+
+Note: You will have to define `HOST1`, `HOST2`, and `HOST3`. Examples could be `$ export HOST1=10.0.0.1` or a resolvable DNS name.
+
+Pay attention to the name parameter in extraArgs. It needs to be unique per host. This example uses infra0, infra1, and infra2, but it can be any valid etcd member name just so long as the names all match in the `initial-cluster` section.
+
+```
+cat << EOF > kubeadmcfg.yaml
+apiVersion: "kubeadm.k8s.io/v1alpha1"
+kind: MasterConfiguration
+etcd:
+    serverCertSANs:
+    - "${HOST1}"
+    - "${HOST2}"
+    - "${HOST3}"
+    peerCertSANs:
+    - "${HOST1}"
+    - "${HOST2}"
+    - "${HOST3}"
+    extraArgs:
+        initial-cluster-token: etcd-cluster-1
+        initial-cluster: infra0=https://${HOST1}:2380,infra1=https://${HOST2}:2380,infra2=https://${HOST3}:2380
+        initial-cluster-state: new
+        name: infra0
+        listen-peer-urls: https://${CURRENT_HOST}:2380
+        listen-client-urls: https://${CURRENT_HOST}:2379
+        advertise-client-urls: https://${CURRENT_HOST}:2379
+        initial-advertise-peer-urls: https://${CURRENT_HOST}:2380
+EOF
+```
+
+## Certificates
+
+### Certificate Authority
+
+Create a CA for the etcd certs. This only has to be done on one node.
```

Comment:
Is it possible to use CA generated outside of kubeadm? It looks like if `/etc/kubernetes/pki/etcd/ca.{crt,key}` exists, kubeadm does not try to create it. Then could we mention that too?

I think this would be useful when using other tools with kubeadm

Reply:
I think that's worth mentioning. I'll add some copy. Thanks for the suggestion.